File: backend/services/options_engine.py
"""
Options Strategy Engine - Generates trade plans

MCP-Ready Functions:
- generate_strategies(): Generate HIGH/MEDIUM/LOW risk strategies
- get_plans(): Retrieve trade plans from database

Each strategy includes Greeks, risk/reward, exit conditions.
"""

from datetime import datetime, timedelta
from typing import List, Optional
from sqlmodel import Session, select

from backend.database import engine
from backend.models import TradePlan
from backend.services import yfinance_service
from backend.services import greeks as greeks_service
from backend.schemas.options import (
    StrategyRequest, StrategyResponse, StrategyPlan, OptionLeg,
    GreeksSnapshot, StrategyType, RiskTier, TradePlanSummary
)

# Current risk-free rate approximation
RISK_FREE_RATE = 0.045  # 4.5%


def generate_strategies(request: StrategyRequest) -> StrategyResponse:
    """
    Generate three options strategies (HIGH/MEDIUM/LOW risk).

    MCP Tool: generate_options_strategies

    Steps:
    1. Fetch options chain
    2. Generate HIGH risk strategy (directional)
    3. Generate MEDIUM risk strategy (spread)
    4. Generate LOW risk strategy (income)
    5. Check capital constraints
    6. Persist plans to database
    7. Return structured response
    """
    symbol = request.symbol.upper()
    capital = request.capital

    # Get current data
    quote = yfinance_service.get_quote(symbol)
    current_price = quote.get("price", 0)

    if not current_price:
        return StrategyResponse(
            symbol=symbol,
            current_price=0,
            capital=capital,
            strategies=[],
            recommendation=f"Unable to get price for {symbol}",
            capital_warnings=["No price data"]
        )

    chain = yfinance_service.get_options_chain(symbol, expiration_index=0)
    chain_monthly = yfinance_service.get_options_chain(symbol, expiration_index=2)

    if "error" in chain:
        return StrategyResponse(
            symbol=symbol,
            current_price=current_price,
            capital=capital,
            strategies=[],
            recommendation=f"Unable to fetch options data: {chain['error']}",
            capital_warnings=["No options data available"]
        )

    strategies = []
    capital_warnings = []

    # HIGH RISK: Weekly directional (nearest expiration)
    high_risk = _create_long_call_plan(symbol, current_price, capital, chain)
    if high_risk:
        strategies.append(high_risk)
        if high_risk.total_debit and high_risk.total_debit > capital:
            capital_warnings.append(f"HIGH RISK: Costs ${high_risk.total_debit:.2f}, exceeds ${capital:.2f} budget")

    # MEDIUM RISK: Bull call spread (~30 days)
    medium_risk = _create_bull_call_spread(symbol, current_price, capital, chain_monthly if "error" not in chain_monthly else chain)
    if medium_risk:
        strategies.append(medium_risk)
        if medium_risk.total_debit and medium_risk.total_debit > capital:
            capital_warnings.append(f"MEDIUM RISK: Costs ${medium_risk.total_debit:.2f}, exceeds ${capital:.2f} budget")

    # LOW RISK: Put credit spread (~30 days)
    low_risk = _create_put_credit_spread(symbol, current_price, capital, chain_monthly if "error" not in chain_monthly else chain)
    if low_risk:
        strategies.append(low_risk)
        if low_risk.collateral_required and low_risk.collateral_required > capital:
            capital_warnings.append(f"LOW RISK: Requires ${low_risk.collateral_required:.2f} collateral, exceeds ${capital:.2f} budget")

    # Build recommendation
    affordable = [s for s in strategies if (s.total_debit or 0) <= capital and (s.collateral_required or 0) <= capital]
    if affordable:
        best = affordable[0]
        recommendation = f"Recommended: {best.strategy_type.value} ({best.risk_tier.value} risk) - fits within ${capital:.2f} budget"
    elif strategies:
        recommendation = f"WARNING: No strategies fit within ${capital:.2f} budget. Consider higher capital or different stock."
    else:
        recommendation = "No strategies could be generated. Check options availability."

    # Persist plans
    with Session(engine) as session:
        for strategy in strategies:
            plan = TradePlan(
                analysis_id=request.analysis_id,
                symbol=symbol,
                strategy_type=strategy.strategy_type.value,
                risk_tier=strategy.risk_tier.value,
                status="PLANNED",
                entry_price=current_price,
                strike_price=strategy.legs[0].strike if strategy.legs else None,
                strike_price_2=strategy.legs[1].strike if len(strategy.legs) > 1 else None,
                expiration_date=strategy.legs[0].expiration if strategy.legs else None,
                contracts=strategy.legs[0].contracts if strategy.legs else 1,
                premium_paid=strategy.total_debit,
                premium_received=strategy.total_credit,
                delta=strategy.legs[0].greeks.delta if strategy.legs and strategy.legs[0].greeks else None,
                implied_volatility=chain.get("atm_call", {}).get("implied_volatility") if chain.get("atm_call") else None,
                max_profit=strategy.max_profit,
                max_loss=strategy.max_loss,
                breakeven_price=strategy.breakeven,
                risk_reward_ratio=strategy.risk_reward_ratio,
                win_probability=strategy.win_probability,
                profit_target_pct=strategy.profit_target_pct,
                stop_loss_pct=strategy.stop_loss_pct,
                time_stop_days=strategy.time_stop_days,
                notes=strategy.rationale
            )
            session.add(plan)
        session.commit()

    return StrategyResponse(
        symbol=symbol,
        current_price=current_price,
        capital=capital,
        strategies=strategies,
        recommendation=recommendation,
        capital_warnings=capital_warnings
    )


def get_plans(symbol: Optional[str] = None, status: Optional[str] = None) -> List[TradePlanSummary]:
    """
    Retrieve trade plans from database.

    MCP Tool: list_trade_plans
    """
    with Session(engine) as session:
        statement = select(TradePlan)

        if symbol:
            statement = statement.where(TradePlan.symbol == symbol.upper())
        if status:
            statement = statement.where(TradePlan.status == status.upper())

        statement = statement.order_by(TradePlan.created_at.desc())
        plans = session.exec(statement).all()

        return [
            TradePlanSummary(
                id=p.id,
                symbol=p.symbol,
                strategy_type=p.strategy_type,
                risk_tier=p.risk_tier,
                status=p.status,
                strike=p.strike_price,
                expiration=p.expiration_date,
                max_profit=p.max_profit,
                max_loss=p.max_loss,
                win_probability=p.win_probability,
                created_at=p.created_at
            )
            for p in plans
        ]


def _create_long_call_plan(
    symbol: str,
    spot: float,
    capital: float,
    chain: dict
) -> Optional[StrategyPlan]:
    """Create HIGH RISK long call strategy."""
    calls = chain.get("calls", [])
    expiration = chain.get("selected_expiration")

    if not calls or not expiration:
        return None

    # Find call ~5% OTM
    target_strike = spot * 1.05

    best_call = None
    for call in calls:
        if call["strike"] >= target_strike and call.get("ask") and call["ask"] > 0:
            best_call = call
            break

    if not best_call:
        # Fallback to ATM
        atm_call = chain.get("atm_call")
        if atm_call and atm_call.get("ask"):
            best_call = atm_call
        else:
            return None

    strike = best_call["strike"]
    premium = best_call.get("ask") or best_call.get("last_price", 0)
    iv = best_call.get("implied_volatility", 30) / 100

    try:
        exp_date = datetime.strptime(expiration, "%Y-%m-%d")
    except:
        exp_date = datetime.now() + timedelta(days=7)

    days_to_exp = max((exp_date - datetime.now()).days, 1)
    time_to_expiry = days_to_exp / 365

    greeks_result = greeks_service.calculate_greeks(
        spot=spot,
        strike=strike,
        time_to_expiry=time_to_expiry,
        volatility=iv,
        risk_free_rate=RISK_FREE_RATE,
        option_type="call"
    )

    total_cost = premium * 100
    max_loss = total_cost
    breakeven = strike + premium

    win_prob = greeks_service.estimate_win_probability(
        spot=spot,
        breakeven=breakeven,
        time_to_expiry=time_to_expiry,
        volatility=iv,
        option_type="call"
    )

    warnings = []
    move_needed = (breakeven - spot) / spot * 100
    if move_needed > 10:
        warnings.append(f"Needs +{move_needed:.1f}% move to breakeven")
    if iv > 0.5:
        warnings.append(f"IV is {iv*100:.0f}% - options are expensive")

    return StrategyPlan(
        strategy_type=StrategyType.LONG_CALL,
        risk_tier=RiskTier.HIGH,
        legs=[
            OptionLeg(
                action="BUY",
                option_type="CALL",
                strike=strike,
                expiration=exp_date,
                premium=premium,
                contracts=1,
                greeks=GreeksSnapshot(
                    delta=greeks_result.delta,
                    gamma=greeks_result.gamma,
                    theta=greeks_result.theta,
                    vega=greeks_result.vega
                )
            )
        ],
        total_debit=total_cost,
        max_profit=999999,
        max_loss=max_loss,
        breakeven=breakeven,
        risk_reward_ratio=0,
        win_probability=win_prob,
        profit_target_pct=50,
        stop_loss_pct=50,
        time_stop_days=max(1, days_to_exp - 1),
        rationale=f"Directional bullish bet. Delta {greeks_result.delta:.2f}. Needs {move_needed:.1f}% move.",
        warnings=warnings
    )


def _create_bull_call_spread(
    symbol: str,
    spot: float,
    capital: float,
    chain: dict
) -> Optional[StrategyPlan]:
    """Create MEDIUM RISK bull call spread."""
    calls = chain.get("calls", [])
    expiration = chain.get("selected_expiration")
    atm_strike = chain.get("atm_strike")

    if not calls or not expiration or not atm_strike:
        return None

    buy_call = None
    for call in calls:
        if call["strike"] == atm_strike and call.get("ask"):
            buy_call = call
            break

    if not buy_call:
        return None

    target_sell_strike = atm_strike * 1.05
    sell_call = None
    for call in calls:
        if call["strike"] >= target_sell_strike and call.get("bid") and call["bid"] > 0:
            sell_call = call
            break

    if not sell_call:
        return None

    buy_premium = buy_call.get("ask", 0)
    sell_premium = sell_call.get("bid", 0)
    net_debit = (buy_premium - sell_premium) * 100

    spread_width = sell_call["strike"] - buy_call["strike"]
    max_profit = (spread_width * 100) - net_debit
    max_loss = net_debit
    breakeven = buy_call["strike"] + (buy_premium - sell_premium)

    try:
        exp_date = datetime.strptime(expiration, "%Y-%m-%d")
    except:
        exp_date = datetime.now() + timedelta(days=30)

    days_to_exp = max((exp_date - datetime.now()).days, 1)
    time_to_expiry = days_to_exp / 365
    iv = buy_call.get("implied_volatility", 30) / 100

    buy_greeks = greeks_service.calculate_greeks(spot, buy_call["strike"], time_to_expiry, iv, RISK_FREE_RATE, "call")
    sell_greeks = greeks_service.calculate_greeks(spot, sell_call["strike"], time_to_expiry, iv, RISK_FREE_RATE, "call")

    spread_delta = buy_greeks.delta - sell_greeks.delta
    risk_reward = max_profit / max_loss if max_loss > 0 else 0

    win_prob = greeks_service.estimate_win_probability(spot, breakeven, time_to_expiry, iv, "call")

    warnings = []
    if net_debit > capital:
        warnings.append(f"Cost ${net_debit:.2f} exceeds budget")

    return StrategyPlan(
        strategy_type=StrategyType.BULL_CALL_SPREAD,
        risk_tier=RiskTier.MEDIUM,
        legs=[
            OptionLeg(
                action="BUY",
                option_type="CALL",
                strike=buy_call["strike"],
                expiration=exp_date,
                premium=buy_premium,
                contracts=1,
                greeks=GreeksSnapshot(
                    delta=buy_greeks.delta,
                    gamma=buy_greeks.gamma,
                    theta=buy_greeks.theta,
                    vega=buy_greeks.vega
                )
            ),
            OptionLeg(
                action="SELL",
                option_type="CALL",
                strike=sell_call["strike"],
                expiration=exp_date,
                premium=sell_premium,
                contracts=1,
                greeks=GreeksSnapshot(
                    delta=sell_greeks.delta,
                    gamma=sell_greeks.gamma,
                    theta=sell_greeks.theta,
                    vega=sell_greeks.vega
                )
            )
        ],
        total_debit=net_debit,
        max_profit=max_profit,
        max_loss=max_loss,
        breakeven=breakeven,
        risk_reward_ratio=round(risk_reward, 2),
        win_probability=win_prob,
        profit_target_pct=50,
        stop_loss_pct=40,
        time_stop_days=max(3, days_to_exp - 5),
        rationale=f"Defined risk spread. Net delta {spread_delta:.2f}. R:R {risk_reward:.1f}:1.",
        warnings=warnings
    )


def _create_put_credit_spread(
    symbol: str,
    spot: float,
    capital: float,
    chain: dict
) -> Optional[StrategyPlan]:
    """Create LOW RISK put credit spread (income strategy)."""
    puts = chain.get("puts", [])
    expiration = chain.get("selected_expiration")

    if not puts or not expiration:
        return None

    target_sell_strike = spot * 0.95
    sell_put = None
    for put in reversed(puts):
        if put["strike"] <= target_sell_strike and put.get("bid") and put["bid"] > 0:
            sell_put = put
            break

    if not sell_put:
        return None

    target_buy_strike = sell_put["strike"] * 0.95
    buy_put = None
    for put in reversed(puts):
        if put["strike"] <= target_buy_strike and put.get("ask"):
            buy_put = put
            break

    if not buy_put:
        return None

    sell_premium = sell_put.get("bid", 0)
    buy_premium = buy_put.get("ask", 0)
    net_credit = (sell_premium - buy_premium) * 100

    spread_width = sell_put["strike"] - buy_put["strike"]
    collateral = spread_width * 100
    max_profit = net_credit
    max_loss = collateral - net_credit
    breakeven = sell_put["strike"] - (sell_premium - buy_premium)

    try:
        exp_date = datetime.strptime(expiration, "%Y-%m-%d")
    except:
        exp_date = datetime.now() + timedelta(days=30)

    days_to_exp = max((exp_date - datetime.now()).days, 1)
    time_to_expiry = days_to_exp / 365
    iv = sell_put.get("implied_volatility", 30) / 100

    sell_greeks = greeks_service.calculate_greeks(spot, sell_put["strike"], time_to_expiry, iv, RISK_FREE_RATE, "put")

    risk_reward = max_profit / max_loss if max_loss > 0 else 0

    win_prob = 100 - greeks_service.estimate_win_probability(spot, sell_put["strike"], time_to_expiry, iv, "put")

    warnings = []
    if collateral > capital:
        warnings.append(f"Requires ${collateral:.2f} collateral - exceeds budget")

    cushion_pct = (spot - sell_put["strike"]) / spot * 100

    return StrategyPlan(
        strategy_type=StrategyType.PUT_CREDIT_SPREAD,
        risk_tier=RiskTier.LOW,
        legs=[
            OptionLeg(
                action="SELL",
                option_type="PUT",
                strike=sell_put["strike"],
                expiration=exp_date,
                premium=sell_premium,
                contracts=1,
                greeks=GreeksSnapshot(
                    delta=sell_greeks.delta,
                    gamma=sell_greeks.gamma,
                    theta=-sell_greeks.theta,
                    vega=-sell_greeks.vega
                )
            ),
            OptionLeg(
                action="BUY",
                option_type="PUT",
                strike=buy_put["strike"],
                expiration=exp_date,
                premium=buy_premium,
                contracts=1,
                greeks=None
            )
        ],
        total_credit=net_credit,
        collateral_required=collateral,
        max_profit=max_profit,
        max_loss=max_loss,
        breakeven=breakeven,
        risk_reward_ratio=round(risk_reward, 2),
        win_probability=win_prob,
        profit_target_pct=50,
        stop_loss_pct=100,
        time_stop_days=max(3, days_to_exp - 5),
        rationale=f"Income strategy. {cushion_pct:.1f}% cushion. Collect ${net_credit:.2f} credit.",
        warnings=warnings
    )

File: apps/api/src/services/options-engine.ts
/**
 * Options Strategy Engine - Generates trade plans
 *
 * Generates HIGH/MEDIUM/LOW risk strategies with Greeks and exit conditions.
 */

import { eq, desc, and } from 'drizzle-orm';
import { db, schema } from '../db/client.js';
import { yahooFinance } from './yahoo-finance.js';
import { calculateGreeks, estimateWinProbability } from '@stock-researcher/shared';
import type {
  StrategyResponse,
  StrategyPlan,
  OptionLeg,
  GreeksSnapshot,
  TradePlanSummary,
} from '@stock-researcher/shared';

// Current risk-free rate approximation
const RISK_FREE_RATE = 0.045; // 4.5%

interface StrategyRequest {
  symbol: string;
  capital: number;
  analysis_id: number | null;
}

interface OptionsChain {
  current_price: number;
  selected_expiration: string;
  atm_strike: number;
  atm_call: FormattedOption | null;
  calls: FormattedOption[];
  puts: FormattedOption[];
  error?: string;
}

interface FormattedOption {
  contract_symbol: string | null;
  strike: number;
  last_price: number | null;
  bid: number | null;
  ask: number | null;
  volume: number | null;
  open_interest: number | null;
  implied_volatility: number;
  in_the_money: boolean;
}

/**
 * Generate three options strategies (HIGH/MEDIUM/LOW risk).
 */
async function generateStrategies(request: StrategyRequest): Promise<StrategyResponse> {
  const symbol = request.symbol.toUpperCase();
  const capital = request.capital;

  // Get current data
  const quote = await yahooFinance.getQuote(symbol);
  const currentPrice = quote.price;

  if (!currentPrice) {
    return {
      symbol,
      current_price: 0,
      capital,
      strategies: [],
      recommendation: `Unable to get price for ${symbol}`,
      capital_warnings: ['No price data'],
    };
  }

  const chain = await yahooFinance.getOptionsChain(symbol, 0);
  const chainMonthly = await yahooFinance.getOptionsChain(symbol, 2);

  if (chain.error) {
    return {
      symbol,
      current_price: currentPrice,
      capital,
      strategies: [],
      recommendation: `Unable to fetch options data: ${chain.error}`,
      capital_warnings: ['No options data available'],
    };
  }

  const strategies: StrategyPlan[] = [];
  const capitalWarnings: string[] = [];

  // HIGH RISK: Weekly directional (nearest expiration)
  const highRisk = createLongCallPlan(symbol, currentPrice, capital, chain);
  if (highRisk) {
    strategies.push(highRisk);
    if (highRisk.total_debit && highRisk.total_debit > capital) {
      capitalWarnings.push(`HIGH RISK: Costs $${highRisk.total_debit.toFixed(2)}, exceeds $${capital.toFixed(2)} budget`);
    }
  }

  // MEDIUM RISK: Bull call spread (~30 days)
  const mediumChain = chainMonthly.error ? chain : chainMonthly;
  const mediumRisk = createBullCallSpread(symbol, currentPrice, capital, mediumChain);
  if (mediumRisk) {
    strategies.push(mediumRisk);
    if (mediumRisk.total_debit && mediumRisk.total_debit > capital) {
      capitalWarnings.push(`MEDIUM RISK: Costs $${mediumRisk.total_debit.toFixed(2)}, exceeds $${capital.toFixed(2)} budget`);
    }
  }

  // LOW RISK: Put credit spread (~30 days)
  const lowRisk = createPutCreditSpread(symbol, currentPrice, capital, mediumChain);
  if (lowRisk) {
    strategies.push(lowRisk);
    if (lowRisk.collateral_required && lowRisk.collateral_required > capital) {
      capitalWarnings.push(`LOW RISK: Requires $${lowRisk.collateral_required.toFixed(2)} collateral, exceeds $${capital.toFixed(2)} budget`);
    }
  }

  // Build recommendation
  const affordable = strategies.filter(
    (s) => (s.total_debit || 0) <= capital && (s.collateral_required || 0) <= capital
  );

  let recommendation: string;
  if (affordable.length > 0) {
    const best = affordable[0];
    recommendation = `Recommended: ${best.strategy_type} (${best.risk_tier} risk) - fits within $${capital.toFixed(2)} budget`;
  } else if (strategies.length > 0) {
    recommendation = `WARNING: No strategies fit within $${capital.toFixed(2)} budget. Consider higher capital or different stock.`;
  } else {
    recommendation = 'No strategies could be generated. Check options availability.';
  }

  // Persist plans
  for (const strategy of strategies) {
    await db.insert(schema.tradePlans).values({
      analysisId: request.analysis_id,
      symbol,
      createdAt: new Date(),
      strategyType: strategy.strategy_type,
      riskTier: strategy.risk_tier,
      status: 'PLANNED',
      entryPrice: currentPrice,
      strikePrice: strategy.legs[0]?.strike ?? null,
      strikePrice2: strategy.legs[1]?.strike ?? null,
      expirationDate: strategy.legs[0]?.expiration ?? null,
      contracts: strategy.legs[0]?.contracts ?? 1,
      premiumPaid: strategy.total_debit ?? null,
      premiumReceived: strategy.total_credit ?? null,
      delta: strategy.legs[0]?.greeks?.delta ?? null,
      impliedVolatility: chain.atm_call?.implied_volatility ?? null,
      maxProfit: strategy.max_profit,
      maxLoss: strategy.max_loss,
      breakevenPrice: strategy.breakeven,
      riskRewardRatio: strategy.risk_reward_ratio,
      winProbability: strategy.win_probability,
      profitTargetPct: strategy.profit_target_pct,
      stopLossPct: strategy.stop_loss_pct,
      timeStopDays: strategy.time_stop_days,
      notes: strategy.rationale,
    });
  }

  return {
    symbol,
    current_price: currentPrice,
    capital,
    strategies,
    recommendation,
    capital_warnings: capitalWarnings,
  };
}

/**
 * Retrieve trade plans from database.
 */
async function getPlans(symbol?: string | null, status?: string | null): Promise<TradePlanSummary[]> {
  const conditions = [];
  if (symbol) {
    conditions.push(eq(schema.tradePlans.symbol, symbol.toUpperCase()));
  }
  if (status) {
    conditions.push(eq(schema.tradePlans.status, status.toUpperCase()));
  }

  const plans = await db.query.tradePlans.findMany({
    where: conditions.length > 0 ? and(...conditions) : undefined,
    orderBy: [desc(schema.tradePlans.createdAt)],
  });

  return plans.map((p) => ({
    id: p.id,
    symbol: p.symbol,
    strategy_type: p.strategyType,
    risk_tier: p.riskTier,
    status: p.status,
    strike: p.strikePrice,
    expiration: p.expirationDate,
    max_profit: p.maxProfit,
    max_loss: p.maxLoss,
    win_probability: p.winProbability,
    created_at: p.createdAt,
  }));
}

/**
 * Create HIGH RISK long call strategy.
 */
function createLongCallPlan(
  symbol: string,
  spot: number,
  capital: number,
  chain: OptionsChain
): StrategyPlan | null {
  const calls = chain.calls;
  const expiration = chain.selected_expiration;

  if (!calls.length || !expiration) {
    return null;
  }

  // Find call ~5% OTM
  const targetStrike = spot * 1.05;

  let bestCall: FormattedOption | null = null;
  for (const call of calls) {
    if (call.strike >= targetStrike && call.ask && call.ask > 0) {
      bestCall = call;
      break;
    }
  }

  if (!bestCall) {
    // Fallback to ATM
    bestCall = chain.atm_call;
    if (!bestCall || !bestCall.ask) {
      return null;
    }
  }

  const strike = bestCall.strike;
  const premium = bestCall.ask || bestCall.last_price || 0;
  const iv = (bestCall.implied_volatility || 30) / 100;

  const expDate = new Date(expiration);
  const daysToExp = Math.max(Math.ceil((expDate.getTime() - Date.now()) / (1000 * 60 * 60 * 24)), 1);
  const timeToExpiry = daysToExp / 365;

  const greeksResult = calculateGreeks(spot, strike, timeToExpiry, iv, RISK_FREE_RATE, 'call');

  const totalCost = premium * 100;
  const maxLoss = totalCost;
  const breakeven = strike + premium;

  const winProb = estimateWinProbability(spot, breakeven, timeToExpiry, iv, 'call');

  const warnings: string[] = [];
  const moveNeeded = ((breakeven - spot) / spot) * 100;
  if (moveNeeded > 10) {
    warnings.push(`Needs +${moveNeeded.toFixed(1)}% move to breakeven`);
  }
  if (iv > 0.5) {
    warnings.push(`IV is ${(iv * 100).toFixed(0)}% - options are expensive`);
  }

  const greeks: GreeksSnapshot = {
    delta: greeksResult.delta,
    gamma: greeksResult.gamma,
    theta: greeksResult.theta,
    vega: greeksResult.vega,
  };

  const leg: OptionLeg = {
    action: 'BUY',
    option_type: 'CALL',
    strike,
    expiration: expDate,
    premium,
    contracts: 1,
    greeks,
  };

  return {
    strategy_type: 'LONG_CALL',
    risk_tier: 'HIGH',
    legs: [leg],
    total_debit: totalCost,
    max_profit: 999999,
    max_loss: maxLoss,
    breakeven,
    risk_reward_ratio: 0,
    win_probability: winProb,
    profit_target_pct: 50,
    stop_loss_pct: 50,
    time_stop_days: Math.max(1, daysToExp - 1),
    rationale: `Directional bullish bet. Delta ${greeksResult.delta.toFixed(2)}. Needs ${moveNeeded.toFixed(1)}% move.`,
    warnings,
  };
}

/**
 * Create MEDIUM RISK bull call spread.
 */
function createBullCallSpread(
  symbol: string,
  spot: number,
  capital: number,
  chain: OptionsChain
): StrategyPlan | null {
  const calls = chain.calls;
  const expiration = chain.selected_expiration;
  const atmStrike = chain.atm_strike;

  if (!calls.length || !expiration || !atmStrike) {
    return null;
  }

  // Find buy call at ATM
  let buyCall: FormattedOption | null = null;
  for (const call of calls) {
    if (call.strike === atmStrike && call.ask) {
      buyCall = call;
      break;
    }
  }

  if (!buyCall) {
    return null;
  }

  // Find sell call ~5% OTM
  const targetSellStrike = atmStrike * 1.05;
  let sellCall: FormattedOption | null = null;
  for (const call of calls) {
    if (call.strike >= targetSellStrike && call.bid && call.bid > 0) {
      sellCall = call;
      break;
    }
  }

  if (!sellCall) {
    return null;
  }

  const buyPremium = buyCall.ask || 0;
  const sellPremium = sellCall.bid || 0;
  const netDebit = (buyPremium - sellPremium) * 100;

  const spreadWidth = sellCall.strike - buyCall.strike;
  const maxProfit = spreadWidth * 100 - netDebit;
  const maxLoss = netDebit;
  const breakeven = buyCall.strike + (buyPremium - sellPremium);

  const expDate = new Date(expiration);
  const daysToExp = Math.max(Math.ceil((expDate.getTime() - Date.now()) / (1000 * 60 * 60 * 24)), 1);
  const timeToExpiry = daysToExp / 365;
  const iv = (buyCall.implied_volatility || 30) / 100;

  const buyGreeks = calculateGreeks(spot, buyCall.strike, timeToExpiry, iv, RISK_FREE_RATE, 'call');
  const sellGreeks = calculateGreeks(spot, sellCall.strike, timeToExpiry, iv, RISK_FREE_RATE, 'call');

  const spreadDelta = buyGreeks.delta - sellGreeks.delta;
  const riskReward = maxLoss > 0 ? maxProfit / maxLoss : 0;

  const winProb = estimateWinProbability(spot, breakeven, timeToExpiry, iv, 'call');

  const warnings: string[] = [];
  if (netDebit > capital) {
    warnings.push(`Cost $${netDebit.toFixed(2)} exceeds budget`);
  }

  const buyLeg: OptionLeg = {
    action: 'BUY',
    option_type: 'CALL',
    strike: buyCall.strike,
    expiration: expDate,
    premium: buyPremium,
    contracts: 1,
    greeks: {
      delta: buyGreeks.delta,
      gamma: buyGreeks.gamma,
      theta: buyGreeks.theta,
      vega: buyGreeks.vega,
    },
  };

  const sellLeg: OptionLeg = {
    action: 'SELL',
    option_type: 'CALL',
    strike: sellCall.strike,
    expiration: expDate,
    premium: sellPremium,
    contracts: 1,
    greeks: {
      delta: sellGreeks.delta,
      gamma: sellGreeks.gamma,
      theta: sellGreeks.theta,
      vega: sellGreeks.vega,
    },
  };

  return {
    strategy_type: 'BULL_CALL_SPREAD',
    risk_tier: 'MEDIUM',
    legs: [buyLeg, sellLeg],
    total_debit: netDebit,
    max_profit: maxProfit,
    max_loss: maxLoss,
    breakeven,
    risk_reward_ratio: Number(riskReward.toFixed(2)),
    win_probability: winProb,
    profit_target_pct: 50,
    stop_loss_pct: 40,
    time_stop_days: Math.max(3, daysToExp - 5),
    rationale: `Defined risk spread. Net delta ${spreadDelta.toFixed(2)}. R:R ${riskReward.toFixed(1)}:1.`,
    warnings,
  };
}

/**
 * Create LOW RISK put credit spread (income strategy).
 */
function createPutCreditSpread(
  symbol: string,
  spot: number,
  capital: number,
  chain: OptionsChain
): StrategyPlan | null {
  const puts = chain.puts;
  const expiration = chain.selected_expiration;

  if (!puts.length || !expiration) {
    return null;
  }

  // Find sell put ~5% OTM
  const targetSellStrike = spot * 0.95;
  let sellPut: FormattedOption | null = null;
  for (let i = puts.length - 1; i >= 0; i--) {
    const put = puts[i];
    if (put.strike <= targetSellStrike && put.bid && put.bid > 0) {
      sellPut = put;
      break;
    }
  }

  if (!sellPut) {
    return null;
  }

  // Find buy put ~5% below sell strike
  const targetBuyStrike = sellPut.strike * 0.95;
  let buyPut: FormattedOption | null = null;
  for (let i = puts.length - 1; i >= 0; i--) {
    const put = puts[i];
    if (put.strike <= targetBuyStrike && put.ask) {
      buyPut = put;
      break;
    }
  }

  if (!buyPut) {
    return null;
  }

  const sellPremium = sellPut.bid || 0;
  const buyPremium = buyPut.ask || 0;
  const netCredit = (sellPremium - buyPremium) * 100;

  const spreadWidth = sellPut.strike - buyPut.strike;
  const collateral = spreadWidth * 100;
  const maxProfit = netCredit;
  const maxLoss = collateral - netCredit;
  const breakeven = sellPut.strike - (sellPremium - buyPremium);

  const expDate = new Date(expiration);
  const daysToExp = Math.max(Math.ceil((expDate.getTime() - Date.now()) / (1000 * 60 * 60 * 24)), 1);
  const timeToExpiry = daysToExp / 365;
  const iv = (sellPut.implied_volatility || 30) / 100;

  const sellGreeks = calculateGreeks(spot, sellPut.strike, timeToExpiry, iv, RISK_FREE_RATE, 'put');

  const riskReward = maxLoss > 0 ? maxProfit / maxLoss : 0;

  // Win probability for credit spread = probability stock stays above breakeven
  const winProb = 100 - estimateWinProbability(spot, sellPut.strike, timeToExpiry, iv, 'put');

  const warnings: string[] = [];
  if (collateral > capital) {
    warnings.push(`Requires $${collateral.toFixed(2)} collateral - exceeds budget`);
  }

  const cushionPct = ((spot - sellPut.strike) / spot) * 100;

  const sellLeg: OptionLeg = {
    action: 'SELL',
    option_type: 'PUT',
    strike: sellPut.strike,
    expiration: expDate,
    premium: sellPremium,
    contracts: 1,
    greeks: {
      delta: sellGreeks.delta,
      gamma: sellGreeks.gamma,
      theta: -sellGreeks.theta,
      vega: -sellGreeks.vega,
    },
  };

  const buyLeg: OptionLeg = {
    action: 'BUY',
    option_type: 'PUT',
    strike: buyPut.strike,
    expiration: expDate,
    premium: buyPremium,
    contracts: 1,
    greeks: null,
  };

  return {
    strategy_type: 'PUT_CREDIT_SPREAD',
    risk_tier: 'LOW',
    legs: [sellLeg, buyLeg],
    total_credit: netCredit,
    collateral_required: collateral,
    max_profit: maxProfit,
    max_loss: maxLoss,
    breakeven,
    risk_reward_ratio: Number(riskReward.toFixed(2)),
    win_probability: winProb,
    profit_target_pct: 50,
    stop_loss_pct: 100,
    time_stop_days: Math.max(3, daysToExp - 5),
    rationale: `Income strategy. ${cushionPct.toFixed(1)}% cushion. Collect $${netCredit.toFixed(2)} credit.`,
    warnings,
  };
}

// Export as module
export const optionsEngine = {
  generateStrategies,
  getPlans,
};
